# syntax-transform/3.1/0021-kernel-match.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, `Kernel#=~` always returns `false`. In Ruby versions 1.9 through
# 3.1, `Kernel#=~` always returns `nil`. In Ruby 3.2 and later, there is no
# longer a `Kernel#=~` method, although there is still a `NilClass#=~` method
# that always returns `nil`. So this code snippet prints `false` in Ruby 1.8,
# `nil` in Ruby 1.9 through 3.1 and throws an error in later versions.
#
#   p 0 =~ nil
#
# This code snippet does the same thing except in Ruby versions later than 3.1,
# where it prints `nil` instead of erroring.
#
#   p nil =~ 0

--- a/object.c
+++ b/object.c
@@ -1318,7 +1318,7 @@ nil_inspect(VALUE obj)
 static VALUE
 nil_match(VALUE obj1, VALUE obj2)
 {
-    return Qnil;
+    return mkxp_ec_is_syntax_transform_active(1, 8, -1) ? Qfalse : Qnil;
 }
 
 /***********************************************************************
@@ -1506,6 +1506,9 @@ rb_false(VALUE obj)
 static VALUE
 rb_obj_match(VALUE obj1, VALUE obj2)
 {
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        return Qfalse;
+    }
     if (rb_warning_category_enabled_p(RB_WARN_CATEGORY_DEPRECATED)) {
         rb_category_warn(RB_WARN_CATEGORY_DEPRECATED, "deprecated Object#=~ is called on %"PRIsVALUE
                 "; it always returns nil", rb_obj_class(obj1));
